docs/design-documents/20221024-stream-inspector.md

```diff
@@ -0,0 +1,110 @@
+   time the inspection started. Historical data is not required.
+6. The inspector should have a minimal footprint on the resource usage (CPU, memory etc.)
+7. The inspector should have an insignificant impact on a pipeline's performance (ideally, no impact at all).
+8. The inspector data should be available through:
+    * the HTTP API,
+    * the gRPC API and
+    * the UI.
+9. Multiple users should be able to inspect the same pipeline component at the same time. Every user's session is
+   independent.
+10. The inspector should stop publishing the data to a client, if the client appears to be idle/crashed.
+
+## Implementation
+
+Implementations will generally use one of two approaches: pull based and push based.
+
+* In a **pull based** implementation, a client (HTTP client, gRPC client, UI) would be periodically checking for
+  inspection data.
+  * A client would need to store the "position" of the previous inspection, to get newer data.
+  * This would require inspection state management in Conduit for each session.
+  * Inspection data would come in with a delay.
+* In a **push based** implementation, Conduit would be pushing the inspection data to a client (HTTP client, gRPC
+  client, UI).
+  * A client would not need to store the "position" of the previous inspection, to get newer data.
+  * Inspection state management in Conduit would be minimal.
+  * Inspection data would come in almost real-time.
+  * Inspecting the data may require additional tools (depends on the actual implementation)
+
+From what we know so far, **the push based approach has more advantages and is easier to work with, so that's the 
+approach chosen here**. Concrete implementation options are discussed below.
+
+For context: gRPC is the main API for Conduit. The HTTP API is generated using [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway).
+
+### API
+Inspecting a pipeline component is triggered with a gRPC/HTTP API request. If the pipeline component cannot be inspected
+for any reason (e.g. inspection not supported, component not found), an error is returned.
+
+Inspectable pipeline components themselves should expose an `Inspect` method, for example
+```go
+func(s Source) Inspect() chan Record 
+```
+(As a return value, we may use a special `struct` instead of `chan Record` to more easily propagate events, such as 
+inspection done.)
+
+The API layer will use this method to get the records it needs to stream to the client which requested an inspection.
+
+For any given pipeline component, only one gRPC/HTTP API method is required, one which starts an inspection (i.e. 
+sending of data to a client).
+
+Three methods are required in total:
+1. One to inspect a connector (there's a single endpoint for source and destination connectors)
+2. One to inspect records coming into a processor
+3. One to inspect records coming out of a processor
```

Comment:
I'll supply another option - we could have one endpoint (e.g. `/pipeline/{id}/inspect`) that takes arguments describing which entity to inspect (e.g. `{"entity":"processor","id":"my-processor"}`). Didn't think through if it's better or not 🙂

Reply:
I like the idea! It would help with the direction, and IMO is better than having different endpoints for different directions.

When it comes to having a single endpoint for all: I'm viewing the connectors and processors as the actual resources, and "inspect" as the action taken on them. So at this moment, I find it a bit more natural to have `/connectors/{id}/inspect` then `/pipeline/{id}/inspect`.

Reply:
You're right, endpoints on the entities probably make more sense. My design likely doesn't even work because the request opens a websocket, so it needs to be a `GET` request.

We need to see if we can send the direction as a URL parameter and if that can be retrieved in the gRPC endpoint. Otherwise we'll have to resort to another option (Header? Separate endpoints? Initial message after the WS is opened that specifies the direction?). No need to figure this out here, it can be left for the implementation if you ask me ;)